app/controllers/image_controller.rb

```diff
@@ -364,62 +410,96 @@ def process_image(args, upload)
     end
   end
 
+  public
+
+  ##############################################################################
+
   # Form for editing date/license/notes on an image.
   # Linked from: show_image/original
   # Inputs: params[:id] (image)
   #   params[:comment][:summary]
   #   params[:comment][:comment]
   # Outputs: @image, @licenses
   def edit_image
-    pass_query_params
-    return unless (@image = find_or_goto_index(Image, params[:id].to_s))
+    return unless image_exists
 
-    @licenses = License.current_names_and_ids(@image.license)
-    if !check_permission!(@image)
-      redirect_with_query(action: "show_image", id: @image)
-    elsif request.method != "POST"
-      init_project_vars_for_add_or_edit(@image)
+    set_licenses_ivar
+    check_image_permission
+    if request.method == "POST"
+      update_image
     else
-      @image.attributes = whitelisted_image_params
-      xargs = {}
-      xargs[:set_date] = @image.when if @image.when_changed?
-      xargs[:set_notes] = @image.notes if @image.notes_changed?
-      if @image.copyright_holder_changed?
-        xargs[:set_copyright_holder] = @image.copyright_holder
-      end
-      if @image.original_name_changed?
-        xargs[:set_original_name] = @image.original_name
-      end
-      xargs[:set_license] = @image.license if @image.license_id_changed?
-      done = false
-      if xargs.empty?
-        if update_projects(@image, params[:project])
-          flash_notice(:runtime_image_edit_success.t(id: @image.id))
-        else
-          flash_notice(:runtime_no_changes.t)
-        end
-        done = true
-      elsif !@image.save
-        flash_object_errors(@image)
-      else
-        xargs[:id] = @image
-        @image.log_update
-        flash_notice(:runtime_image_edit_success.t(id: @image.id))
-        update_projects(@image, params[:project])
-        done = true
-      end
-      if done
-        redirect_with_query(action: "show_image", id: @image.id)
-      else
-        init_project_vars_for_reload(@image)
-      end
+      init_project_vars_for_add_or_edit(@image)
     end
   end
 
   ##############################################################################
 
   private
 
+  def update_image
+    return unless image_exists
+
+    set_licenses_ivar
+    check_image_permission
+
+    @image.attributes = whitelisted_image_params
+    xargs = set_xargs
+
+    done = figure_out_if_done(xargs)
+    if done
+      redirect_with_query(action: "show_image", id: @image.id)
+    else
+      init_project_vars_for_reload(@image)
+    end
+  end
+
+  def set_licenses_ivar
```

Comment:
This is clear... but just throwing it out there, would it be even better to have the caller set @licenses and then name the method something that actually reveals what it's setting @licenses to, such as "current_names_and_ids" or something like that?  Even _more_ transparent, and just as short / concise.